confluent_kafka/schema_registry/protobuf.py

```diff
@@ -0,0 +1,437 @@
+            # Write the magic byte and schema ID in network byte order (big endian)
+            fo.write(struct.pack('>bI', _MAGIC_BYTE, self._schema_id))
+            # write the record index to the buffer
+            self._encode_index(fo, self.msg_index)
+            # write the record itself
+            fo.write(msg.SerializeToString())
+            return fo.getvalue()
+
+
+class ProtobufDeserializer(object):
+    """
+    ProtobufDeserializer decodes bytes written in the Schema Registry
+    Protobuf format to an object.
+
+    Args:
+        descriptor (Descriptor): Protobuf Message Descriptor.
+
+    .. _Schema definition:
+        https://json-schema.org/understanding-json-schema/reference/generic.html
+
+    """
+    def __init__(self, descriptor):
+        self.msg_index = _create_msg_index(descriptor)
+        self.msg_class = MessageFactory().GetPrototype(descriptor)
+
+    @staticmethod
+    def _decode_uvarint(buf):
+        """
+        Decodes a single uvarint from a buffer.
+
+        Args:
+            - buf (BytesIO): buffer to read from
+
+        Returns:
+            int: decoded uvarint
+
+        Raises:
+            EOFError: if buffer is empty
+
+        """
+        value = 0
+        shift = 0
+        while True:
+            i = ProtobufDeserializer._read_byte(buf)
+
+            value |= (i & 0x7f) << shift
+            shift += 7
+            if not (i & 0x80):
+                break
+
+        return value
+
+    @staticmethod
+    def _read_byte(buf):
+        """
+        Returns int representation for a byte.
+
+        Args:
+            buf (BytesIO): buffer to read from
+
+        .. _ord:
+            https://docs.python.org/2/library/functions.html#ord
+        """
+        i = buf.read(1)
+        if i == b'':
+            raise EOFError("Unexpected EOF while reading index")
+        return ord(i)
+
+    @staticmethod
+    def _decode_index(buf, expect):
+        """
+        Extracts message index from Schema Registry Protobuf formatted bytes.
+
+        Args:
+            buf (BytesIO): byte buffer
+
+            expect ([int]): expected message index
+
+        Returns:
+            [int]: Message index. Not currently used.
+
+        Raises:
+            SerializationError: If message indexes do not match
+
+        """
+        expect_size = expect[0]
+        size = ProtobufDeserializer._decode_uvarint(buf)
+        # fail fast, this is guaranteed not to be our expected message type.
+        if expect_size != size:
+            raise SerializationError("Message index mismatch. The consumed"
+                                     " Message does not match this"
+                                     " Deserializer's Message type")
+
+        msg_index = [size]
+        for _ in range(size):
+            msg_index.append(ProtobufDeserializer._decode_uvarint(buf))
+
+        if msg_index != expect:
+            raise SerializationError("Message index mismatch. The consumed"
```

Comment:
this is fine - they don't need to match, they just need to be compatible. `ParseFromString` will tell you that. you should remove the `expect` checking stuff.

Reply:
This will need to be addressed in the multi-type serializer implementation to come in a follow up PR. I'll try to get it in this afternoon.